plutus-core/plutus-ir/src/PlutusIR/Compiler/Types.hs

```diff
@@ -49,19 +49,21 @@ instance PLC.HasTypeCheckConfig (PirTCConfig uni fun) uni fun where
     typeCheckConfig = pirConfigTCConfig
 
 data CompilationOpts a = CompilationOpts {
-    _coOptimize                   :: Bool
-    , _coPedantic                 :: Bool
-    , _coVerbose                  :: Bool
-    , _coDebug                    :: Bool
-    , _coMaxSimplifierIterations  :: Int
+    _coOptimize                       :: Bool
+    , _coPedantic                     :: Bool
+    , _coVerbose                      :: Bool
+    , _coDebug                        :: Bool
+    , _coMaxSimplifierIterations      :: Int
     -- Simplifier passes
-    , _coDoSimplifierUnwrapCancel :: Bool
-    , _coDoSimplifierCaseReduce   :: Bool
-    , _coDoSimplifierBeta         :: Bool
-    , _coDoSimplifierInline       :: Bool
-    , _coInlineHints              :: InlineHints PLC.Name (Provenance a)
-    , _coProfile                  :: Bool
-    , _coRelaxedFloatin           :: Bool
+    , _coDoSimplifierUnwrapCancel     :: Bool
+    , _coDoSimplifierCaseReduce       :: Bool
+    , _coDoSimplifierBeta             :: Bool
+    , _coDoSimplifierInline           :: Bool
+    , _coDoSimplifierEvaluateBuiltins :: Bool
```

Comment:
If the only problematic one is `trace`, can we instead have a flag that controls the evaluation of only `trace`?

In fact I would like a more general flag that controls whether the passes are allowed to change logging. Because CSE and other passes may also change logging.

We can also consider adding optimization levels, where changes to logging is only allowed with `-O2` or above.

Reply:
`trace` isn't the only one. All builtins _can_ emit log messages. And we can't know which ones the users care about really.

> In fact I would like a more general flag that controls whether the passes are allowed to change logging. 

That's what this flag was supposed to be. I'm not sure why I gave it the unnecessarily specific name. It _does_ currently control whether the pass will evaluate _any_ builtin with logging!

> We can also consider adding optimization levels, where changes to logging is only allowed with -O2 or above.

I think the current segmentation makes more sense for us. For normal compilers the tradeoff with higher optimization levels is usually "better code but slower compilation". We don't really care about slower compilation at the moment, so we'd always take the max level if we had them, so it's not so useful. But "does this mess with the effects" _does_ seem like something people will care about.

Reply:
I renamed the option to "preserveLogging`, does that seem more okay?

Reply:
>  For normal compilers the tradeoff with higher optimization levels is usually "better code but slower compilation". 

It's a separate discussion, orthogonal to this PR, but high optimization levels can also mean "risky optimizations", e.g., gcc's loop unrolling which may increase program size.

For us, we can use "risky optimization" imply both "may change logging" (evaluating builtins, CSE) and "may increase cost/size" (float-in, float-out). Having a higher optimization level would be a convenient way to say "just try your best to optimize!"